hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -618,6 +655,37 @@ private CompletableFuture<Message> writeStateMachineData(
     return raftFuture;
   }
 
+  private void validateLongRunningWrite(long currIndex) throws IOException {
+    // get min valid write chunk operation's future context
+    WriteFutureContext writeFutureContext = null;
+    for (long i = writeFutureMinIndex; i < currIndex; ++i) {
+      if (writeChunkFutureMap.containsKey(i)) {
+        writeFutureContext = writeChunkFutureMap.get(i);
+        writeFutureMinIndex = i;
+        break;
+      }
+    }
+    if (null == writeFutureContext) {
+      return;
+    }
+    // validate for timeout in milli second
+    long waitTime = Time.monotonicNow() - writeFutureContext.getStartTime() / 1000000;
+    IOException ex = new StorageContainerException("Write chunk has taken " + waitTime + " crossing threshold "
+        + writeChunkWaitMaxMs + " for groupId " + getGroupId(), ContainerProtos.Result.CONTAINER_INTERNAL_ERROR);
+    if (waitTime > writeChunkWaitMaxMs) {
+      LOG.error("Write chunk has taken {}ms crossing threshold {}ms for groupId {}", waitTime, writeChunkWaitMaxMs,
+          getGroupId());
+      stateMachineHealthy.set(false);
+      writeChunkFutureMap.forEach((key, value) -> {
+        LOG.error("Cancelling write chunk for transaction {}, groupId {}", key, getGroupId());
+        value.getWriteChunkFuture().cancel(true);
+        value.getRaftFuture().completeExceptionally(ex);
```

Comment:
Just cancelling is enough. Interrupted exception will be caught and that flow will take care of the rest